Linux/sgx/test_app/enclave/tests/evp_smx.c

```diff
@@ -36,591 +36,666 @@
-            printf("Error: fail to read public key from the BIO\n");
-            ret = -11;
-            break;
-        }
-        (*public_key)[pub_len] = '\0';
-
-    } while(0);
-
-    // 6. Finalize
-    if (ret != 0)
-    {
-        SAFE_FREE(*private_key, pri_len+1);
-        SAFE_FREE(*public_key, pub_len+1);
-    }	
-	EC_KEY_free(ec_key);
+	int ret = 0;
+	BIO *pri_bio = NULL, *pub_bio = NULL;
+	size_t pri_len = 0, pub_len = 0;
+	EVP_PKEY *evp_pkey = NULL;
+	EVP_PKEY_CTX *evp_pkey_ctx = NULL;
+
+	do {
+#if OPENSSL_VERSION_NUMBER < 0x30000000
+		printf("OpenSSL version: 1.1.1 or lower\n");
+		// 1. Create an EC_GROUP object with a curve specified by SM2 NID
+		EC_GROUP *ec_group = NULL;
+		EC_KEY *ec_key = NULL;
+		ec_group = EC_GROUP_new_by_curve_name(NID_sm2);
+		if (ec_group == NULL) {
+			printf("Error: fail to create an EC_GROUP object for SM2\n");
+			ret = -1;
+			break;
+		}
+		// 2. Create a new EC key
+		ec_key = EC_KEY_new();
+		if (ec_key == NULL) {
+			printf("Error: fail to create a new EC key\n");
+			ret = -2;
+			break;
+		}
+		// 3. Set the new EC key's curve
+		if (EC_KEY_set_group(ec_key, ec_group) != 1) {
+			printf("Error: fail to set the new EC key's curve\n");
+			ret = -3;
+			break;
+		}
+		// 4. Generate key pair based on the curve
+		if (!EC_KEY_generate_key(ec_key)) {
+			printf("Error: fail to generate key pair based on the curve\n");
+			ret = -4;
+			break;
+		}
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000
+		printf("OpenSSL version: 3.0.0 or higher\n");
+		// 1. Create a EVP_PKEY_CTX for SM2
+		evp_pkey_ctx = EVP_PKEY_CTX_new_from_name(NULL, "SM2", NULL);
+		if (evp_pkey_ctx == NULL) {
+			printf("Error: fail to create a EVP_PKEY_CTX for SM2\n");
+			ret = -5;
+			break;
+		}
+		// 2. Initialize the SM2 EVP_PKEY_CTX
+		if (EVP_PKEY_keygen_init(evp_pkey_ctx) != 1) {
+			printf("Error: fail to initialize the SM2 EVP_PKEY_CTX\n");
+			ret = -6;
+			break;
+		}
+		// 3. Set params for the SM2 EVP_PKEY_CTX
+		OSSL_PARAM params[2];
+		params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME, "SM2", 0);
+		params[1] = OSSL_PARAM_construct_end();
+		if (EVP_PKEY_CTX_set_params(evp_pkey_ctx, params) != 1) {
+			printf("Error: fail to set params for SM2 EVP_PKEY_CTX\n");
+			ret = -7;
+			break;
+		}
+		// 4. Generate a SM2 EVP_PKEY by the SM2 EVP_PKEY_CTX
+		if (EVP_PKEY_generate(evp_pkey_ctx, &evp_pkey) != 1) {
+			printf("Error: fail to generate the SM2 EVP_PKEY by SM2 EVP_PKEY_CTX\n");
+			ret = -8;
+			break;
+		}
+#endif
+
+		// 5. Generate SM2 private key
+		pri_bio = BIO_new(BIO_s_mem());
+		if (pri_bio == NULL) {
+			printf("Error: fail to create a BIO for SM2 private key\n");
+			ret = -9;
+			break;
+		}
+#if OPENSSL_VERSION_NUMBER < 0x30000000
+		if (!PEM_write_bio_ECPrivateKey(pri_bio, ec_key, NULL, NULL, 0, NULL, NULL)) {
+			printf("Error: fail to write SM2 private key from ec_key to the BIO\n");
+			ret = -10;
+			break;
+		}
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000
```

Comment:
You can simply use only `#else` here, since it should be enough when it is `#if OPENSSL_VERSION_NUMBER < 0x30000000` above.